src/osd/PeeringState.cc

```diff
@@ -5117,13 +5117,18 @@ PeeringState::Backfilling::react(const DeferBackfill &c)
   ps->state_clear(PG_STATE_BACKFILLING);
   cancel_backfill();
 
-  pl->schedule_event_after(
-    std::make_shared<PGPeeringEvent>(
-      ps->get_osdmap_epoch(),
-      ps->get_osdmap_epoch(),
-      RequestBackfill()),
-    c.delay);
-  return transit<NotBackfilling>();
+  if (ps->needs_backfill()) {
+    pl->schedule_event_after(
+      std::make_shared<PGPeeringEvent>(
+	ps->get_osdmap_epoch(),
+	ps->get_osdmap_epoch(),
+	RequestBackfill()),
+      c.delay);
+    return transit<NotBackfilling>();
+  } else {
+    // raced with MOSDPGBackfill::OP_BACKFILL_FINISH, ignore
+    return discard_event();
+  }
```

Comment:
Can you please explain this?

Reply:
If PeeringState is in Backfilling and the PG doesn't really needs backfill, it must be at the end of the backfilling where all objects are backfilled while the final phase "OP_BACKFILL_FINISH" has not finished.